azavea/echo-locator
# Add translation strings

## Overview

This PR adds translation strings that switch when the translation buttons are clicked. 

### Demo

![Screen Shot 2021-12-13 at 11 16 46 AM](https://user-images.githubusercontent.com/77936689/145848427-e37cea36-d30c-4623-9045-271a580689b1.png)

Translated into Spanish

![Screen Shot 2021-12-13 at 11 16 55 AM](https://user-images.githubusercontent.com/77936689/145848467-303b6947-e880-49aa-b8fd-0426305ec9f8.png)

Translated in Chinese

### Notes

While some of the strings were very straight forward, a lot of them were not. Here is an overview of the problems faced and the decisions made. 

**Missing Translations**
* Missing translations in **Spanish**: "Peak", "Off Peak" (in dock when commuter trains are selected in profile), "Site by" (dock)
* Missing translations in **Chinese**: "New Search" (nav bar), "when" (dock), "Apartments" (nav bar), "Resend invitation email" (edit profile), "Primary" (edit profile)
* Missing translations in **both languages**: "Peak No Express", "Off Peak No Express" (in dock when commuter trains are _not_ selected in profile), "of" (in the dock, e.g. 1-3 of 142), all neighborhood descriptions

**Amplify Code**

There were also elements that had translations but were not incorporated for similar reasons as laid out in [PR 377](https://github.com/azavea/echo-locator/pull/377). As quoted from that PR, the log in components, especially "Sign Out", "Username", "Password", "Forgot your password?", and "Reset password" are not translated "because [Amplify code] apparently uses its own, separate translation machinery. This primarily affects the login flow, and the translation switcher isn't available on the login page anyway, so I didn't try to find a solution. In any event, we won't be able to use Amplify's login components once we switch to Django authentication, so eventually the login components will be migrated to `react-i18next` too."

**Grammar in Dynamic Phrases**

One of the issues I ran into with translating dynamic phrases into Chinese was how we structured sentences. For example, the phrase "Take [bus number] in [number] minutes" in the dock is currently structured with "Take", "in", and "minutes" as separate phrases that get translated separately and pieced together into a sentence along with the dynamic values within the component. There are grammatical issues with some of the existing translations, as well, especially in phrases that have been structured piece-meal. One example is in the "[mode of transportation] from [destination]" phrase in the dock. This particular sentence has grammatical issues at least in Spanish and possibly also in Chinese, again due to how we have chosen to structure that component. These sentences also have an added complexity in that the phrases used to complete the mode of transportation and destination name are connected to other pieces of the app, particularly the Profile page, where one version of the word might make sense but then does not when used in a sentence in another piece of the app. As an example, the app uses the Spanish word "trabajo" in the Profile page destination dropdown menu but that word should become "el trabajo" when used in a sentence and, if the word before it is "de", should become "del trabajo."  A separate task (#389) has been added to explore alternate ways of structuring the problematic phrases to allow for more flexible grammatical rules. 

**Peak/Off Peak/Peak No Express/Off Peak No Express Dropdown Menu**

The Peak/Off Peak/Peak No Express/Off Peak No Express Dropdown Menu has not been translated, as I ran into an issue with how the `Select` element in `form.js` generates the options and sends those options to network state. I tried a few different methods, such as adding the translate method into a mapping function that generates the network options to the Select component and translating within the Select function as the options are generated; however, there is a disconnect between how the option labels need to look to work with `i18next` and the option labels that state expects. Ultimately, for the purpose of this PR, I decided that this should be set aside as its own future task (#391) because we currently do not have translations for any of the options in Spanish and for two of the options in Chinese.

**Maintaining State Across Multiple Languages**

As of right now, the destination label in the "Show Travel Times: from" dropdown menu is always in the language that it was set in from the Profile page and not in the language that is currently active. This hints at a larger issue as to how state understands the same value but in a different language that may be related to the Peak/Off Peak menu issue related above; however, I'm not sure this particular manifestation is a huge issue, as in theory users will not be going back and forth between languages.

## Testing Instructions

 * Either run `./scripts/server` or `cd taui` &#8594; `yarn start`
 * Click on the "Español" button. Check the dock, neighborhood detail view, marker pop ups, and edit profile page to verify that any English has been addressed above.
 * Click on the "中文" button. Check the dock, neighborhood detail view, marker pop ups, and edit profile page to verify that any English has been addressed above.


Resolves #384
